src/harness/reference_models/pre_iap_filtering/pre_iap_util.py

```diff
@@ -0,0 +1,185 @@
+  return pal_frequencies[0]
+
+def getFssNeighboringCbsdsWithGrants(cbsds, fss_record, distance):
+  """Get the list of all CBSDs that lie within 150 KMs of FSS and have at least one grant.
+
+  Args:
+    cbsds :  List of CbsdData dictionaries as defined in the SAS-SAS specification.
+    fss_record: A FSS record dictionary.
+  Returns:
+    List of CBSDs lying within 150 KMs of FSS and having at least one grant.
+  """
+
+  neighboring_cbsds_with_grants = []
+  for cbsd in cbsds:
+    distance_km, _, _ = vincenty.GeodesicDistanceBearing(
+        fss_record['record']['deploymentParam'][0]['installationParam']['latitude'],
+        fss_record['record']['deploymentParam'][0]['installationParam']['longitude'],
+        cbsd['registrationRequest']['installationParam']['latitude'],
+        cbsd['registrationRequest']['installationParam']['longitude'])
+    # Get the list of cbsds that are within 150kms from the FSS entity
+    if distance_km <= distance and cbsd['grants']:
+        neighboring_cbsds_with_grants.append(cbsd)
+  return neighboring_cbsds_with_grants
+
+
+def getFssNeighboringGwbl(gwbl_records, fss_records, distance): 
+  """Get the list of all fss that lie within 150 KMs of GWBL and operating below 3700 MHz
+
+  Args:
+    fss_records:  List of FSS records dictionary.
+    gwbl_records: List of GWBL record dictionary.
+  Returns:
+    List of FSSs lying within 150 KMs of GWBL and operating below 3700000000.
+  """
+
+  list_of_fss_neighboring_gwbl = []
+  for fss_record in fss_records:
+    for gwbl_record in gwbl_records:
+      # Get the distance of the FSS entity from the GWBL polygon
+      distance_km, _, _ = vincenty.GeodesicDistanceBearing(
+          fss_record['record']['deploymentParam'][0]['installationParam']['latitude'],
+          fss_record['record']['deploymentParam'][0]['installationParam']['longitude'],
+          gwbl_record['record']['deploymentParam'][0]['installationParam']['latitude'],
+          gwbl_record['record']['deploymentParam'][0]['installationParam']['longitude']) 
+      # Get the list of FSS entity that are 150kms from the GWBL area
+      if distance_km <= distance and fss_record['record']['deploymentParam'][0]\
+             ['operationParam']['operationFrequencyRange']['highFrequency'] < 3700000000:
+        list_of_fss_neighboring_gwbl.append(fss_record)
+        break
+  return list_of_fss_neighboring_gwbl
+
+
+def getCbsdsNotPartOfPpaCluster(cbsds, ppa_record):
+  """Returns the CBSDs that are not part of a PPA cluster list.
+
+  Args:
+    cbsds : List of CBSDs.
+    ppa_record : A PPA record dictionary.
+  Returns:
+    List of CBSDs that are not part of the PPA cluster list.
+  """
+
+  cbsds_not_part_of_ppa_cluster = []
+  # Compare the list of CBSDs with the PPA cluster list
+  for cbsd in cbsds:
+    for ppa_cluster_cbsd_id in ppa_record['ppaInfo']['cbsdReferenceId']:
+      if cbsd['id'] != ppa_cluster_cbsd_id:
+        cbsds_not_part_of_ppa_cluster.append(cbsd)
```

Comment:
To be reviewed.
we should check that cbsd['id'] is not in the set of ppa_record['ppaInfo']['cbsdReferenceId'], currently we are checking element-wise, which is wrong.

Reply:
Updated the code, reference id is formed with FCC +SN and checked the presence of cbsd reference id in cluster reference ids.